datafusion/expr/src/expr.rs

```diff
@@ -1461,6 +1462,176 @@ impl Expr {
             | Expr::Placeholder(..) => false,
         }
     }
+
+    /// This method hashes the direct content of an expression node without recursing into
+    /// its children. This is useful because in `CommonSubexprEliminate` we can build up
+    /// the deep hash of a node and its descendants during the bottom-up phase of the
+    /// first traversal and so avoid computing the hash of the node and then the hash of
+    /// its descendants separately.
+    ///
+    /// As it is pretty easy to forget changing this method when `Expr` changes the
+    /// implementation doesn't use wildcard patterns (`..`, `_`) to catch changes
+    /// compile time.
```

Comment:
Can we please also note if this method will return the same value as expr.hash for expressions with no children. I don't think it needs to, but it would be good to explicitly document the intent

```suggestion
    /// Hashes the direct content of an `Expr` without recursing into
    /// its children.
    ///
    /// This method is useful to incrementally compute Hashes, such as  in `CommonSubexprEliminate`
    /// which builds a
    /// deep hash of a node and its descendants during the bottom-up phase of the
    /// first traversal and so avoid computing the hash of the node and then the hash of
    /// its descendants separately.
    ///
    /// As it is pretty easy to forget changing this method when `Expr` changes the
    /// implementation doesn't use wildcard patterns (`..`, `_`) to catch changes
    /// compile time.
```

Reply:
Fixed in https://github.com/apache/datafusion/pull/10473/commits/ee61224f1a948c38ac99ae53d9d426bb69da9548.